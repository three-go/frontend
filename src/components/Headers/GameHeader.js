import React from "react";

import { Text, View, StyleSheet, Pressable } from "react-native";
import Icon from "react-native-vector-icons/MaterialCommunityIcons";

import { ChanceIcons, TextTimer } from "..";
import { iconNames, camera, colors, iconSizes, game } from "../../common";

const GameHeader = ({
  status,
  setStatus,
  onPressBack,
  chance,
  score,
  cameraPermissionStatus,
  currentGameKey,
}) => {
  const handleSetStatusDirectionInput = () => {
    setStatus(game.status.directionInput);
  };

  const handleSetStatusPlay = () => {
    setStatus(game.status.play);
  };

  return (
    <View style={styles.container}>
      {status !== game.status.none && (
        <Pressable onPress={onPressBack} style={styles.back}>
          <Icon
            name={iconNames.leftArrow}
            size={iconSizes.headerLeftArrow}
            color={colors.gray}
          />
        </Pressable>
      )}

      {currentGameKey === game.keys[0] && (
        <View>
          {status === game.status.open && (
            <TextTimer
              onTimerEnd={handleSetStatusDirectionInput}
              status={status}
            />
          )}

          {status === game.status.directionInput &&
            cameraPermissionStatus === camera.permissionReady && (
              <TextTimer onTimerEnd={handleSetStatusPlay} status={status} />
            )}

          {status === game.status.play && (
            <Text style={styles.score}>
              SCORE{"\n"}
              {score}
            </Text>
          )}
        </View>
      )}

      {currentGameKey === game.keys[1] && (
        <View style={styles.scoreWrapper}>
          <Icon
            name={iconNames.flag}
            size={iconSizes.headerFlag}
            color={colors.gray}
          />

          <Text style={styles.score}>
            {progressRate.current} / {progressRate.total}
          </Text>
        </View>
      )}

      {status !== game.status.none && <ChanceIcons chance={chance} />}
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flexDirection: "row",
    justifyContent: "space-around",
    alignItems: "center",
    width: "100%",
    height: "100%",
    paddingLeft: 8,
    paddingRight: 8,
    backgroundColor: colors.dark,
    color: colors.ivory,
  },
  back: {
    width: 65,
    height: 40,
  },
  scoreWrapper: {
    flexDirection: "row",
    justifyContent: "center",
    alignItems: "center",
  },
  score: {
    marginLeft: 10,
    color: colors.ivory,
    textAlign: "center",
    fontSize: 24,
    fontWeight: "bold",
  },
});

export default React.memo(GameHeader);
