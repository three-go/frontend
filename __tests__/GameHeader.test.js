import React from "react";

import { render, fireEvent } from "@testing-library/react-native";

import GameHeader from "../src/components/Headers/GameHeader";

jest.mock("react-native-sound", () => {
  return function Test() {
    return {};
  };
});

describe("GameHeader.js", () => {
  const chance = 3;
  const score = 100;
  const cameraPermissionStatus = "READY";

  it("FaceGo before the map is covered", () => {
    const mockOnPressBack = jest.fn();
    const mockSetStatus = jest.fn();
    const gameStatus = "mapOpen";
    const gameType = "faceGo";

    const component = render(
      <GameHeader
        status={gameStatus}
        setStatus={mockSetStatus}
        onPressBack={mockOnPressBack}
        chance={chance}
        score={score}
        cameraPermissionStatus={cameraPermissionStatus}
        currentGameKey={gameType}
      />
    );

    const backButton = component.getByTestId("backIcon");
    fireEvent.press(backButton);

    expect(component.getByText(/맵이 가려지기/i)).toBeTruthy();
    expect(component.getByTestId("backIcon")).toBeTruthy();
    expect(mockOnPressBack.mock.calls.length).toBe(1);
    expect(component.toJSON().children[2].children.length).toBe(3);
  });

  it("FaceGo before the input ends", () => {
    const mockOnPressBack = jest.fn();
    const mockSetStatus = jest.fn();
    const gameStatus = "directionInput";
    const gameType = "faceGo";

    const component = render(
      <GameHeader
        status={gameStatus}
        setStatus={mockSetStatus}
        onPressBack={mockOnPressBack}
        chance={chance}
        score={score}
        cameraPermissionStatus={cameraPermissionStatus}
        currentGameKey={gameType}
      />
    );

    const backButton = component.getByTestId("backIcon");
    fireEvent.press(backButton);

    expect(component.getByText(/입력이 종료되기/i)).toBeTruthy();
    expect(component.getByTestId("backIcon")).toBeTruthy();
    expect(mockOnPressBack.mock.calls.length).toBe(1);
    expect(component.toJSON().children[2].children.length).toBe(3);
  });

  it("FaceGo results for the game", () => {
    const mockOnPressBack = jest.fn();
    const mockSetStatus = jest.fn();
    const gameStatus = "resultPlay";
    const gameType = "faceGo";

    const component = render(
      <GameHeader
        status={gameStatus}
        setStatus={mockSetStatus}
        onPressBack={mockOnPressBack}
        chance={chance}
        score={score}
        cameraPermissionStatus={cameraPermissionStatus}
        currentGameKey={gameType}
      />
    );

    const backButton = component.getByTestId("backIcon");
    fireEvent.press(backButton);

    expect(component.getByText(/SCORE/i)).toBeTruthy();
    expect(component.getByText(new RegExp(score, "i"))).toBeTruthy();
    expect(component.getByTestId("backIcon")).toBeTruthy();
    expect(mockOnPressBack.mock.calls.length).toBe(1);
    expect(component.toJSON().children[2].children.length).toBe(3);
  });

  it("shoutGo play game", () => {
    const mockOnPressBack = jest.fn();
    const mockSetStatus = jest.fn();
    const gameStatus = "resultPlay";
    const gameType = "shoutGo";

    const component = render(
      <GameHeader
        status={gameStatus}
        setStatus={mockSetStatus}
        onPressBack={mockOnPressBack}
        chance={chance}
        score={score}
        cameraPermissionStatus={cameraPermissionStatus}
        currentGameKey={gameType}
      />
    );

    const backButton = component.getByTestId("backIcon");
    fireEvent.press(backButton);

    expect(component.getByText(/SCORE/i)).toBeTruthy();
    expect(component.getByText(new RegExp(score, "i"))).toBeTruthy();
    expect(component.getByTestId("backIcon")).toBeTruthy();
    expect(mockOnPressBack.mock.calls.length).toBe(1);
    expect(component.toJSON().children[2].children.length).toBe(3);
    expect(component.toJSON).toMatchSnapshot();
  });
});
