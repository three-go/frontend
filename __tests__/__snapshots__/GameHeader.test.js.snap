// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameHeader.js shoutGo play game 1`] = `[Function]`;
